File: backend/src/main/java/br/com/calciolari/datahub/imports/infrastructure/interpdv/qrp/EmfPage.java
package br.com.calciolari.datahub.imports.infrastructure.interpdv.qrp;

/**
 * Byte range of one embedded EMF page inside a QRP container.
 * Offsets observed in {@code docs/poc/index.html} ({@code findEmfPages}).
 */
public record EmfPage(int start, int length) {

	public EmfPage {
		if (start < 0) {
			throw new IllegalArgumentException("start must be >= 0");
		}
		if (length <= 0) {
			throw new IllegalArgumentException("length must be > 0");
		}
	}

	public int endExclusive() {
		return Math.addExact(start, length);
	}
}

File: backend/src/Calciolari.DataHub/Imports/Infrastructure/InterPdv/Qrp/EmfPage.cs
namespace Calciolari.DataHub.Imports.Infrastructure.InterPdv.Qrp;

/// <summary>
/// Byte range of one embedded EMF page inside a QRP container.
/// Offsets observed in docs/poc/index.html (findEmfPages).
/// </summary>
public sealed record EmfPage(int Start, int Length)
{
    public int EndExclusive => checked(Start + Length);

    public static EmfPage Create(int start, int length)
    {
        if (start < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(start), "start must be >= 0");
        }

        if (length <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(length), "length must be > 0");
        }

        return new EmfPage(start, length);
    }
}
